django_mongodb/indexes.py

```diff
@@ -0,0 +1,101 @@
+from django.core.exceptions import EmptyResultSet, FullResultSet
+from django.db import NotSupportedError
+from django.db.models import Index
+from django.db.models.expressions import Col
+from django.db.models.fields.related_lookups import In
+from django.db.models.lookups import BuiltinLookup
+from django.db.models.sql.query import Query
+from django.db.models.sql.where import AND, XOR, WhereNode
+
+from .query_utils import process_rhs
+
+mongo_operators_idx = {
+    "exact": "$eq",
+    "gt": "$gt",
+    "gte": "$gte",
+    "lt": "$lt",
+    "lte": "$lte",
+    "in": "$in",
+}
+
+
+def _get_condition_mql(self, model, schema_editor):
+    """Analogous to Index._get_condition_sql()."""
+    query = Query(model=model, alias_cols=False)
+    where = query.build_where(self.condition)
+    compiler = query.get_compiler(connection=schema_editor.connection)
+    return where.as_mql_idx(compiler, schema_editor.connection)
+
+
+def builtin_lookup_idx(self, compiler, connection):
+    if not isinstance(self.lhs, Col):
+        raise NotSupportedError("Expressions as indexes are not supported.")
+    lhs_mql = self.lhs.target.column
+    value = process_rhs(self, compiler, connection)
+    try:
+        operator = mongo_operators_idx[self.lookup_name]
+    except KeyError:
+        raise NotSupportedError(
+            f"MongoDB does not support the '{self.lookup_name}' lookup in indexes."
+        ) from None
+    return {lhs_mql: {operator: value}}
+
+
+def in_idx(self, compiler, connection):
+    if not connection.features.supports_in_index_operator:
+        raise NotSupportedError(
+            f"MongoDB < 6.0 does not support the {self.lookup_name} lookup in indexes."
+        )
+    return builtin_lookup_idx(self, compiler, connection)
+
+
+def where_node_idx(self, compiler, connection):
+    if self.connector == AND:
+        full_needed, empty_needed = len(self.children), 1
+    else:
+        full_needed, empty_needed = 1, len(self.children)
+    if self.connector == AND:
+        operator = "$and"
+    elif self.connector == XOR:
+        raise NotSupportedError("MongoDB does not support the '^' operator lookup in indexes.")
+    else:
+        if not connection.features.supports_in_index_operator:
+            raise NotSupportedError("MongoDB < 6.0 does not support the '|' operator in indexes.")
+        operator = "$or"
+    if self.negated:
+        raise NotSupportedError("MongoDB does not support the '~' operator in indexes.")
+    children_mql = []
+    for child in self.children:
+        try:
+            mql = child.as_mql_idx(compiler, connection)
+        except EmptyResultSet:
+            empty_needed -= 1
+        except FullResultSet:
+            full_needed -= 1
+        else:
+            if mql:
+                children_mql.append(mql)
+            else:
+                full_needed -= 1
+
+    if empty_needed == 0:
```

Comment:
Could you carefully audit these branches? I added `assert False` to them and it doesn't appear test apps `constraints indexes indexes_` covers them. I'm not sure what it would mean to have an index where node that matches nothing or everything. Wouldn't that be kind of pointless? I guess as long as we match Django's behavior, it's okay.

Reply:
Yes, I will review carefully those edge cases and edit code or add test if it is needed.
I think there isn't a test in indexes that covers, I tried to imitate the other where.as_mql.

Reply:
It isn't possible to create a case that covers that given the conditions of the indexes must be a Q expression, I did it in order to emulate the where.as_mql.  This part won't be used and the other try and catch expression also not needed.